cirq-core/cirq/transformers/routing/initial_mapper.py

```diff
@@ -0,0 +1,57 @@
+# Copyright 2022 The Cirq Developers
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from typing import TYPE_CHECKING, Dict
+import abc
+
+from cirq.value import value_equality
+
+if TYPE_CHECKING:
+    import cirq
+
+
+@value_equality
+class AbstractInitialMapper(metaclass=abc.ABCMeta):
+    """Base class for creating custom initial mapping strategies."""
+
+    @abc.abstractmethod
+    def initial_mapping(self) -> Dict['cirq.Qid', 'cirq.Qid']:
+        """Maps the logical qubits of a circuit onto physical qubits on a device.
+        Returns:
+          qubit_map: the initial mapping of logical qubits to physical qubits.
+        """
+
+    def __str__(self) -> str:
+        return f'Logical to physical qubits: {self.initial_mapping()}'
```

Comment:
usually python `__str__` methods are more terse than this. Devs can always add context to a str() message but it's hard to take it away. Usually in Cirq we're pretty picky about the `__repr__` and just use that for the `__str__`